### PR TITLE
Add heartbeat health metric

### PR DESCRIPTION
**Pull request by morrisonlevi**
### Description

This adds a gauge metric to each request called `datadog.tracer.heartbeat` with a default sample rate of 0.1%. The sample rate can be controlled via the `DD_TRACE_HEALTH_METRICS_HEARTBEAT_SAMPLE_RATE` environment variable.

### Readiness checklist
- [ ] Tests added for this feature/bug.

### Reviewer checklist
- [x] Appropriate labels assigned.
- [x] Milestone is set.
- [ ] Changelog has been added to the appropriate release draft. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: src/ext/configuration.h [hunk 1]
```diff
@@ -10,27 +10,28 @@ void ddtrace_initialize_config(TSRMLS_D);
...
-        "reaching this percent threshold of a span buffer will trigger background thread "                \
+#define DD_CONFIGURATION                                                                                             \
+    CHAR(get_dd_agent_host, "DD_AGENT_HOST", "localhost")                                                            \
+    CHAR(get_dd_dogstatsd_port, "DD_DOGSTATSD_PORT", "8125")                                                         \
+    INT(get_dd_trace_agent_port, "DD_TRACE_AGENT_PORT", 8126)                                                        \
+    BOOL(get_dd_trace_debug, "DD_TRACE_DEBUG", FALSE)                                                                \
+    BOOL(get_dd_trace_agent_debug_verbose_curl, "DD_TRACE_AGENT_DEBUG_VERBOSE_CURL", FALSE)                          \
+    BOOL(get_dd_trace_debug_curl_output, "DD_TRACE_DEBUG_CURL_OUTPUT", FALSE)                                        \
+    BOOL(get_dd_trace_heath_metrics_enabled, "DD_TRACE_HEALTH_METRICS_ENABLED", FALSE)                               \
+    DOUBLE(get_dd_trace_heath_metrics_heartbeat_sample_rate, "DD_TRACE_HEALTH_METRICS_HEARTBEAT_SAMPLE_RATE", 0.001) \
```
Comment (morrisonlevi) [inline src/ext/configuration.h]: It's not clear from the diff, but this is the added value; the rest is just reformatting from clang-format.

File: src/ext/configuration.c [hunk 2]
```diff
@@ -20,14 +22,16 @@ void ddtrace_reload_config(TSRMLS_D) {
         free(ddtrace_memoized_configuration.getter_name); \
     }                                                     \
     ddtrace_memoized_configuration.__is_set_##getter_name = FALSE;
-#define INT(getter_name, ...) ddtrace_memoized_configuration.__is_set_##getter_name = FALSE;
 #define BOOL(getter_name, ...) ddtrace_memoized_configuration.__is_set_##getter_name = FALSE;
+#define INT(getter_name, ...) ddtrace_memoized_configuration.__is_set_##getter_name = FALSE;
+#define DOUBLE(getter_name, ...) ddtrace_memoized_configuration.__is_set_##getter_name = FALSE;
```
Comment (SammyK) [inline src/ext/configuration.c]: 🔝

File: src/ext/configuration_render.h [hunk 3]
```diff
@@ -1,5 +1,5 @@
-#ifndef DD_CONFIGURATION_REDER_H
-#define DD_CONFIGURATION_REDER_H
+#ifndef DD_CONFIGURATION_RENDER_H
```
Comment (SammyK) [inline src/ext/configuration_render.h]: Nice catch!

File: src/ext/env_config.c [hunk 4]
```diff
@@ -92,6 +92,36 @@ uint32_t ddtrace_get_uint32_config(char *name, uint32_t def TSRMLS_DC) {
...
 
+double ddtrace_get_double_config(char *name, double def TSRMLS_DC) {
+    char *env = get_local_env_or_sapi_env(name TSRMLS_CC);
+    if (!env) {
+        return def;
+    }
+
+    char *endptr = env;
+
+    // The strtod function is a bit tricky, so I've quoted docs to explain code
```
Comment (SammyK) [inline src/ext/env_config.c]: 💯

File: src/dogstatsd/client.c [hunk 5]
```diff
@@ -116,10 +118,22 @@ dogstatsd_client_status dogstatsd_client_metric_send(
   const char *tags_prefix = (tags_len + const_tags_len > 0) ? "|#" : "";
   const char *tags_separator = (tags_len > 0 && const_tags_len > 0) ? "," : "";
 
-  const char *format = "%s:%s|%s|@%.6f%s%s%s%s";
-  int size = snprintf(client->msg_buffer, client->msg_buffer_len, format, name,
-                      value, typestr, sample_rate, tags_prefix, tags,
-                      tags_separator, client->const_tags);
+  const char *format;
+  int size;
+  /* Omit the sample rate iff it is 1.0; a sample rate of 1.000000 causes issues
```
Comment (SammyK) [inline src/dogstatsd/client.c]: Tiny type-o, `iff`. Otherwise, thanks for the comment! :)

Comment (morrisonlevi) [reply src/dogstatsd/client.c]: [You're one of today's lucky 10000](https://xkcd.com/1053/)! It's a short-hand for "if and only if" in mathematics :)

**Pull request merged**